src/Forms/DiscoverReaders.jsx

```diff
@@ -12,34 +12,63 @@ class DiscoverReaders extends React.Component {
     super(props);
 
     this.state = {
-      discoveryInProgress: false
+      discoveryInProgress: false,
+      requestInProgress: false,
+      discoveryWasCancelled: false
     };
   }
 
   onTriggerDiscoverReaders = async () => {
-    this.setState({ discoveryInProgress: true });
+    // user clicked discovery button ("cancel") while already discovering
```

Comment:
Hrm, do you think it'd be cleaner to have a separate cancel button with its own handler? That might let us get away with no needing `discoveryWasCancelled`

Reply:
This is a nice idea. I'm curious how we then deal with the in-flight discovery request that cannot be cancelled as it's a promise. Right now I'm using this property to also [block showing the list of readers](https://github.com/stripe/stripe-terminal-js-demo/pull/80/files#diff-72557b194881fa6a81542f6f515df2b8R71) when the request completes and populates the `readers` list.

Reply:
Ah, yeah you'd have to push the cancel state boolean up to `MainPage.jsx` to hold onto it there and then use that to prevent the returned readers from being written into state. It could still get a little gross and if some one wanted to bang on the button the could break it, buuuuut this is just the sample app so I don't want to overthink it. You could probably get away with keeping the button disabled until the previous request returns if you wanted to be safe about it.

Reply:
you make an excellent point about flooding requests through hitting the button. I think your idea about disabling the button until the cancellation completes is a good one that I'll implement!